app/src/services/version.js

```diff
@@ -0,0 +1,99 @@
+const { v4: uuidv4, NIL: SYSTEM_USER} = require('uuid');
+
+const { Version } = require('../db/models');
+
+/**
+ * The Version DB Service
+ */
+const service = {
+  /**
+   * @function create
+   * Saves a version of an object
+   * @param {object[]} data array of object attributes and `versionId` from S3 object
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<object>} The result of running the insert operation
+   * @throws The error encountered upon db transaction failure
+   */
+  create: async (data = {}, currentUserId = SYSTEM_USER, etrx = undefined) => {
+
+    if (!data.id) {
+      throw new Error('Invalid objectId supplied');
+    }
+
+    let trx;
+    try {
+      trx = etrx ? etrx : await Version.startTransaction();
+
+      const version = {
+        id: uuidv4(),
+        versionId: data.VersionId,
+        objectId: data.id,
+        createdBy: currentUserId,
+        mimeType: data.mimeType,
+        originalName: data.originalName
+      };
+      const response = await Version.query().insert(version);
+      if (!etrx) await trx.commit();
+      return response;
+    } catch (err) {
+      if (!etrx && trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  /**
+   * @function createVersionOfObjects
+   * Saves a version of each object in an array
+   * @param {object[]} data array of objects each with an `objectId` and version data
+   * @param {object} [etrx=undefined] An optional Objection Transaction object
+   * @returns {Promise<object>} The result of calling the create() method for all objects
+   * @throws The error encountered upon db transaction failure
+   */
+  createVersionOfObjects: async (data = [], userId, etrx = undefined) => {
+    let trx;
+    try {
+      trx = etrx ? etrx : await Version.startTransaction();
+
+      // wait for all create version calls
+      const response =  Promise.all(data.map(async (object) => {
+        await service.create(object, userId, trx);
+      }));
```

Comment:
I recommend changing this to use the insert([ array ]).returns("*") pattern instead as this would reduce the volume of explicit SQL calls done within the transaction. For example, if you were uploading 3 files, using the current promise/map paradigm would yield 3 separate insert statements, whereas doing it with the insert array instead will only yield one insert statement. The latter would be more efficient in both network traffic as well as sql query operation.

Reply:
now that i dont have the isLatest stuff happening, there's no need to call create() for each object. Done multi-row insert as suggested here.